### PR TITLE
Don't give up on retry too early

### PR DESCRIPTION
**Pull request by CyberDem0n**
Fixes https://github.com/zalando/patroni/issues/1195

### CODE REVIEW

File: patroni/dcs/etcd.py [hunk 1]
```diff
@@ -190,31 +216,51 @@ def api_execute(self, path, method, params=None, timeout=None):
...
+                while not response:
+                    response = self._do_http_request(request_executor, method, self._base_uri + path, **kwargs)
+
+                    if response is False:
+                        if not retry:
+                            raise etcd.EtcdException
+                        some_request_failed = True
+                if some_request_failed:
+                    self._refresh_machines_cache()
+                break
```
Comment (sdudoladov) [inline patroni/dcs/etcd.py]: does this `break` mark the happy case when a request retry succeeded ?

File: patroni/dcs/etcd.py [hunk 2]
```diff
@@ -190,31 +216,51 @@ def api_execute(self, path, method, params=None, timeout=None):
...
+            except etcd.EtcdConnectionFailed as e:
+                if isinstance(e, etcd.EtcdWatchTimedOut):
```
Comment (sdudoladov) [inline patroni/dcs/etcd.py]: Does `EtcdWatchTimedOut` always imply a etcd problem thus justifying `self._next_server()` or can there the case where the watched key simply does not change and that triggers the timeout ?

File: patroni/dcs/etcd.py [hunk 3]
```diff
@@ -190,31 +216,51 @@ def api_execute(self, path, method, params=None, timeout=None):
...
-                raise
+        while True:
+            try:
+                some_request_failed = False
+                while not response:
+                    response = self._do_http_request(request_executor, method, self._base_uri + path, **kwargs)
+
+                    if response is False:
+                        if not retry:
+                            raise etcd.EtcdException
```
Comment (sdudoladov) [inline patroni/dcs/etcd.py]: please add the message to the exception

File: patroni/dcs/etcd.py (hunk 1 above)
Comment (CyberDem0n) [reply patroni/dcs/etcd.py]: Yes, added the `if response` there

File: patroni/dcs/etcd.py (hunk 2 above)
Comment (CyberDem0n) [reply patroni/dcs/etcd.py]: I moved it to the `_do_http_request` and left a comment

**Comment by CyberDem0n** [comment]
1. In case of broken network connectivity Patroni was quickly going over list of etcd nodes and giving up much earlier than retry_timeout.
2. The `EtcdLeaderElectionInProgress` is handled by the retry [object](https://github.com/zalando/patroni/pull/1245/files#diff-b19d6b14b30078b89e33a7666e672cd2L318-L319).

**Comment by sdudoladov** [comment]
:+1:

**Pull request merged**